source/opt/loop_peeling.cpp

```diff
@@ -0,0 +1,343 @@
+  ir::BasicBlock* cloned_header = cloned_loop_->GetHeaderBlock();
+  pre_header->ForEachSuccessorLabel(
+      [cloned_header](uint32_t* succ) { *succ = cloned_header->id(); });
+
+  // Update cfg.
+  cfg.RemoveEdge(pre_header->id(), loop_->GetHeaderBlock()->id());
+  cloned_loop_->SetPreHeaderBlock(pre_header);
+  loop_->SetPreHeaderBlock(nullptr);
+
+  // When cloning the loop, we didn't cloned the merge block, so currently
+  // |cloned_loop_| shares the same block as |loop_|.
+  // We mutate all branches from |cloned_loop_| block to |loop_|'s merge into a
+  // branch to |loop_|'s header (so header will also be the merge of
+  // |cloned_loop_|).
+  uint32_t cloned_loop_exit = 0;
+  for (uint32_t pred_id : cfg.preds(loop_->GetMergeBlock()->id())) {
+    if (loop_->IsInsideLoop(pred_id)) continue;
+    ir::BasicBlock* bb = cfg.block(pred_id);
+    assert(cloned_loop_exit == 0 && "The loop has multiple exits.");
+    cloned_loop_exit = bb->id();
+    bb->ForEachSuccessorLabel([this](uint32_t* succ) {
+      if (*succ == loop_->GetMergeBlock()->id())
+        *succ = loop_->GetHeaderBlock()->id();
+    });
+  }
+
+  // Update cfg.
+  cfg.RemoveNonExistingEdges(loop_->GetMergeBlock()->id());
+  cfg.AddEdge(cloned_loop_exit, loop_->GetHeaderBlock()->id());
+
+  // Set the merge block of the cloned loop as the original loop's header block.
+  cloned_loop_->SetMergeBlock(loop_->GetHeaderBlock());
+
+  // Patch the phi of the header.
+  loop_->GetHeaderBlock()->ForEachPhiInst(
+      [cloned_loop_exit, def_use_mgr, this](ir::Instruction* phi) {
+        for (uint32_t i = 1; i < phi->NumInOperands(); i += 2) {
+          if (!loop_->IsInsideLoop(phi->GetSingleWordInOperand(i))) {
+            phi->SetInOperand(i, {cloned_loop_exit});
+            def_use_mgr->AnalyzeInstUse(phi);
+            return;
+          }
+        }
+      });
+
+  // Patch the iterating value initializers of the original loop using the
+  // corresponding cloned loop exit values. Connects the cloned loop iterating
+  // values to the original loop. This make sure that the initial value of the
+  // second loop starts with the last value of the first loop.
+  //
+  // For example, loops like:
+  //
+  // int z = 0;
+  // for (int i = 0; i++ < M; i += cst1) {
+  //   if (cond)
+  //     z += cst2;
+  // }
+  //
+  // Will become:
+  //
+  // int z = 0;
+  // int i = 0;
+  // for (; i++ < M; i += cst1) {
+  //   if (cond)
+  //     z += cst2;
+  // }
+  // for (; i++ < M; i += cst1) {
+  //   if (cond)
+  //     z += cst2;
+  // }
+  loop_->GetHeaderBlock()->ForEachPhiInst(
+      [&clone_results, def_use_mgr, this](ir::Instruction* phi) {
+        for (uint32_t i = 0; i < phi->NumInOperands(); i += 2) {
+          uint32_t pred_id = phi->GetSingleWordInOperand(i + 1);
+          if (!loop_->IsInsideLoop(pred_id)) {
+            phi->SetInOperand(
+                i, {clone_results.value_map_.at(
+                       exit_value_.at(phi->result_id())->result_id())});
```

Comment:
I'm not sure to understand your comment, `clone_results.value_map_.at` returns the corresponding instruction in the cloned loop. What this line is doing is:
  * Taking the instruction holding the update of the phi in the original loop;
  * Get the clone instruction in the cloned loop and use the id for the phi operand.

Reply:
Sorry, I missed that.  Now it makes it easier to explain what I want.  Why don't you simply do:

``
phi->SetInOperand(i, {clone_results.value_map_.at(phi->GetSingleWordInOperand(i))});
```

Are we always guarenteed that the value will be in `clone_results.value_map_`?  What if we have

```
bool b = false;
for( ... ) {
   b = true;
}
```

Reply:
Values in `exit_value_` are guaranteed to be defined in the loop, the map is set in `LoopPeeling::GetIteratingExitValues()`. I think it is yet another baldly named variable.

What it holds is the SSA value that represent the value that each iterators (phi instructions in the header blocks) holds when we exit the loop. i.e.:
```
 int z = 0;
 for (int i = 0; i++ < M; i += cst1) {
    if (cond)
      z += cst2;
 }
```
In this example, `exit_value_` will hold i and z values when we exit the loop. But it may not be the phi node it-self (that happens in do-while loops). With this map, the peeling do something like:
```
 int z = 0;
 for (int i = 0; i++ < 2; i += cst1) {
    if (cond)
      z += cst2;
 }
 int z2 = z;
 for (int i2 = i; i2++ < M; i2 += cst1) {
    if (cond)
      z2 += cst2;
 }
```

What I call "loop iterators" may also have a better name, I don't use "induction variable" because it is too restrictive (z in my example is not an induction variable).

Reply:
As I continue to look at it, I now have a better understanding.  I think the determination of  `exit_value_` can be simpler.

Suppose we have:

```
// bb2
for( ... ) {
   // bb 3
  %4 = OpPhi( %5, bb2, %6, bb3 );
  %6 = ...;
}
```
When looking at the OpPhi, we see that `%6` is the value that comes along the back-edge because `bb3` is the only bb mentioned in the OpPhi that is in the loop.  So the exit value should be `%6`.  So you replace `%5` with the mapped value for %6.  No need to check if you have a do-while loop, or look at dominators.

The other case we need to handle is when the value is not defined in the loop.  Say you have something like:

```
bool first = true;
for( ... ) {
   if (first) {}
   first = false;
}
```

Reply:
Replacing the value %5 by the mapped %6 is only true if your loop exit is also your back-edge (exit the loop or branch to the header), so do-while like loops.

GLSL, HLSL will lower a for loop to something that looks like:
```
// bb2

header:
%4 = OpPhi( %5, bb2, %6, bb3 )
%should_exit = <evaluate exit condition>
cond br %should_exit, %bb_exit, %bb3

bb3:
%6 = ...
br %header

// bb_exit
```
In this case, %4 this value to use.

Reply:
Right, the mapped value does not necessarily dominate the exit, so it cannot always be used.  Okay.  I'm still not convinced the current mechanism is correct.  I'll tyry to convince myself.

Reply:
Side note (that may help you), I try to be conservative, so the utility will not run if there is case like:
```
for (int i = 0; i + 1 < N; i++) {}
for (int i = 0; ++i < N;) {}
```
As depending on the optimization, it may lead to some ambiguous update of the phi (see the second test case). Here, (I think that) a common sub-expression elimination can yield the same code for both cases, so I bail out here. If loops are rotated, this is easier to handle (it is reduced to your initial comment).

Note that
```
for (int i = 0; i < N + 1; i++) {}
```
is fine (the update is not on the IV)